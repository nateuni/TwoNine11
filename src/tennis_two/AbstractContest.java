package tennis_two;

import static util.Contract.require;

import util.Two;

public abstract class AbstractContest implements Contest
{
    
    public final Player             firstServer;
    
    /**
     * Constructs the Contest and initialises the server
     * @param firstServer - passed in to initialise the ivar firstServer.
     */
    public AbstractContest (Player firstServer)
    {
        this.firstServer = firstServer;
    }
    
    /* (non-Javadoc)
     * @see tennis_two.Contest#players()
     */
    public Two <Player> players ()
    {
        return TennisFactory.instance ().players ();
    }
    
    /* (non-Javadoc)
     * @see tennis_two.Contest#score(tennis_two.Player)
     */
    public abstract int score (Player player);
    

    public boolean isNew ()
    {
        return score (server ()) == 0 && score (receiver ()) == 0;
    }
    
    /* (non-Javadoc)
     * @see tennis_two.Contest#scoreAsString(tennis_two.Player)
     */
    public String scoreAsString (Player player)
    {
        return Integer.toString (score (player));
    }
    
    /* (non-Javadoc)
     * @see tennis_two.Contest#isOver()
     */
    public abstract boolean isOver ();
    
    /* (non-Javadoc)
     * @see tennis_two.Contest#winner()
     */
    public Player winner ()
    {
        require (isOver (), "isOver");
        if (score (players ()._1) > score (players ()._2))
            return players ()._1;
        else
            return players ()._2;
    }
    
    
    /* (non-Javadoc)
     * @see tennis_two.Contest#loser()
     */
    public Player loser ()
    {
        return players ().other (winner ());
    }
    
    /* (non-Javadoc)
     * @see tennis_two.Contest#server()
     */
    public Player server ()
    {
        return firstServer;
    }
    
    /* (non-Javadoc)
     * @see tennis_two.Contest#receiver()
     */
    public Player receiver ()
    {
        return players ().other (server ());
    }
    
    /* (non-Javadoc)
     * @see tennis_two.Contest#play()
     */
    public abstract void play ();

    /* (non-Javadoc)
     * @see tennis_two.Contest#display()
     */
    public void display ()
    {
        if (TennisFactory.instance ().displayIsEnabled())
            overridableDisplay ();
    }

    /**
     * Calls helper method displayPlayerScore with each player, which results in the player's score being output to console. 
     */
    protected void overridableDisplay ()
    {
        displayPlayerScore (players()._1);
        displayPlayerScore (players()._2);
    }

    /**
     * Prints the passed players score to screen as a formatted string. 
     * @param player - takes the player who's details will be printed to screen. 
     */
    protected void displayPlayerScore (Player player)
    {
        System.out.format ("%-15s%s%n", player, scoreAsString (player));
    }
    
}