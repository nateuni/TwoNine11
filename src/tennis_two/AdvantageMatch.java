package tennis_two;

import java.util.ArrayList;

/**
 * An advantage match is made up of advantage sets. 
 * Typically a match will be best of 3 or best of 5 sets, as determined by the maxSetCount argument of the factory method.
 * The first server in the match serves in the first game of the first set. 
 * The serve then alternates between the players in successive games.
 */
public class AdvantageMatch extends SerialContest implements Match {
	
	/**
	 * This array list is used to keep Contest pointers for later use in printing contest scores.
	 */
	private static ArrayList<Contest> matchList;
	
	/**
	 * Used to aggregate the total games played, which then will define the next server.
	 */
	private static int totalGamesPlayed;

	protected AdvantageMatch(Player firstServer, int targetScore, int advantage) {
		super(firstServer, targetScore, advantage);
		matchList = new ArrayList<Contest>();
	}
	
	/* (non-Javadoc)
	 * @see tennis_two.Contest#getContestName()
	 */
	@Override
	public String getContestName() {
			return "AdvantageMatch";
	}
	
	/**
	 * Adds the Contest reference to a list and aggregates total games played. 
	 */
	protected void addContestToMatchList() {
		matchList.add(this.currentSubContest);
		totalGamesPlayed += this.currentSubContest.score(this.currentSubContest.server()) + this.currentSubContest.score(this.currentSubContest.receiver());
	}
	
	
	/**
	 * Instantiates a sub-Contest and then plays it.
	 * After Contest has been played it is added a list 
	 * Scores of the current Contest are then printed.
	 */
	@Override
	protected void playSubContest () {
		currentSubContest = makeSubContest ();
		currentSubContest.play ();
		
		addContestToMatchList();

		onSubContestOver ();
	}
	
	/**
	 * After one Set has been played with the first Server of the Match the server is then 
	 * alternated based on if the total number of games played are odd or even.
	 */
	@Override
	protected Contest makeSubContest() {
		if(this.currentSubContest == null){
			return TennisFactory.instance().makeAdvantageSet(firstServer);
		} else if (totalGamesPlayed % 2 == 0) {
			return TennisFactory.instance().makeAdvantageSet(this.server());
		} else {
			return TennisFactory.instance().makeAdvantageSet(this.receiver());
		}
	}
	
	/**
	 * Method is overridden to include to see if the definition of a the end of Tie-Break Match has been met.
	 */
	@Override
	protected void overridableDisplay () {	
		System.out.format("%-15s", this.players()._1);
		for(Contest c : matchList){
			System.out.format("  %s", c.scoreAsString(c.players()._1));
		}
		
		System.out.println();
		
		System.out.format("%-15s", this.players()._2);
		for(Contest c : matchList){
			System.out.format("  %s", c.scoreAsString(c.players()._2));
		}
		
		System.out.println();
    }
}
