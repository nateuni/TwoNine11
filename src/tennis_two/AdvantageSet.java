package tennis_two;

/**
 * An advantage set is won by the first player to get to 6 games with an advantage of 2.
 */
public class AdvantageSet extends SerialContest implements Set {
	
	protected AdvantageSet(Player firstServer, int targetScore, int advantage) {
		super(firstServer, targetScore, advantage);
	}

	/* (non-Javadoc)
	 * @see tennis_two.Contest#getContestName()
	 */
	@Override
	public String getContestName() {
			return "AdvantageSet";
	}
	
	/**
	 * Plays an initial SinglesGame before then alternating servers. 
	 */
	@Override
	protected Contest makeSubContest() {
		if(this.currentSubContest == null) {
			return TennisFactory.instance().makeSinglesGame(firstServer);
		} else if(this.server() == this.currentSubContest.server()) {
			return TennisFactory.instance().makeSinglesGame(this.receiver());
		} else {
			return TennisFactory.instance().makeSinglesGame(this.server());
		}
	}
}