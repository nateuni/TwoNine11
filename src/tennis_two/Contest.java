package tennis_two;

import util.Two;

public interface Contest
{
	/**
     * Used to access the pair of players for the associated instance. 
     * @return Returns "Two" which contains a pair of players
     */
    Two <Player> players ();
    
    /**
     * Returns the score of the associated player as an int. 
     * This methods takes an argument of type Player and then returns that player's score as an int. 
     * @param player - takes an argument of a particular player.
     * @return Returns the passed players score as an int value.
     */
    int score (Player player);
    
    /**
     * Returns the score of the associated player as a String. 
     * This methods takes an argument of type Player and then returns that player's score as a String. 
     * @param player - takes an argument of a particular player.
     * @return Returns the passed players score as a String.
     */
    String scoreAsString (Player player);
    
    /**
     * Checks if the contest instance is new
     * @return returns true if both players scores are equal to 0 or else returns false.
     */
    boolean isNew ();
    
    /**
     * Checks if the contest instance is over
     * @return returns true if the restraints of the particular contest instance have been met.  
     */
    boolean isOver ();
    
    /**
     * Returns the winner
     * @return Returns winning player as type Player.
     */
    Player winner ();
    
    /**
     * Returns the loser
     * @return Returns loosing player as type Player.
     */
    Player loser ();
    
    /**
     * Returns the server of the contest instance. 
     * @return Returns serving player as type Player.
     */
    Player server ();
    
    /**
     * Returns the receiver of the contest instance. . 
     * @return Returns receiving player as type Player.
     */
    Player receiver ();
    
    /**
     * Get the current contest's name
     * @return Returns the contest's name as a String.
     */
    String getContestName ();
    
    /**
     * Display the scores of the associated Contest
     */
    void display ();
    
    /**
     * Initiates the playing of the current contest
     */
    void play ();
    
}
