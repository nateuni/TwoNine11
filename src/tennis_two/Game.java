package tennis_two;

public interface Game extends Contest
{}
