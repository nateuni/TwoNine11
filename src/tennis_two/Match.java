package tennis_two;

public interface Match extends Contest
{}
