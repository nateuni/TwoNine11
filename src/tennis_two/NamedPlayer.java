package tennis_two;

public class NamedPlayer implements Player
{
    
    final public String name;
    
    public NamedPlayer (String name)
    {
        this.name = name;
    }
    
    @Override
    public String toString ()
    {
        return name;
    }
    
    @Override
    public boolean equals (Object other)
    {
        return other instanceof NamedPlayer
               && ((NamedPlayer) other).name.equals (name);
    }
    
}
