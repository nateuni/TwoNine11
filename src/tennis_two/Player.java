package tennis_two;

public interface Player
{}
