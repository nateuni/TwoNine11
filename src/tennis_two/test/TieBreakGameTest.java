package tennis_two.test;

import java.util.Arrays;
import java.util.Collection;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameters;

import tennis_two.Contest;

@RunWith (Parameterized.class)
public class TieBreakGameTest extends AbstractTest
{
    
    @Parameters
    public static Collection <Object []> data ()
    {
        return Arrays.asList (
                new Object [] [] {
                    {111, 3, 7},
                    {444, 7, 5},
                    {1234567, 6, 8},
                    {9876, 10, 8}
                }
            );
    }
    
    public TieBreakGameTest (long seed, int score_1, int score_2)
    {
        super (seed, score_1, score_2);
    }
    
    @Override
    protected Contest contest ()
    {
        return factory.makeTieBreakGame (federer);
    }
}
