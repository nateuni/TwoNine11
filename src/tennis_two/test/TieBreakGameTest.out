TieBreakGame starts
Federer serves and wins
Federer        1
Nadal          0
Nadal serves and loses
Federer        2
Nadal          0
Nadal serves and wins
Federer        2
Nadal          1
Federer serves and loses
Federer        2
Nadal          2
Federer serves and wins
Federer        3
Nadal          2
Nadal serves and wins
Federer        3
Nadal          3
Nadal serves and wins
Federer        3
Nadal          4
Federer serves and loses
Federer        3
Nadal          5
Federer serves and loses
Federer        3
Nadal          6
Nadal serves and wins
Federer        3
Nadal          7
Nadal wins TieBreakGame

TieBreakGame starts
Federer serves and wins
Federer        1
Nadal          0
Nadal serves and wins
Federer        1
Nadal          1
Nadal serves and loses
Federer        2
Nadal          1
Federer serves and loses
Federer        2
Nadal          2
Federer serves and wins
Federer        3
Nadal          2
Nadal serves and wins
Federer        3
Nadal          3
Nadal serves and wins
Federer        3
Nadal          4
Federer serves and loses
Federer        3
Nadal          5
Federer serves and wins
Federer        4
Nadal          5
Nadal serves and loses
Federer        5
Nadal          5
Nadal serves and loses
Federer        6
Nadal          5
Federer serves and wins
Federer        7
Nadal          5
Federer wins TieBreakGame

TieBreakGame starts
Federer serves and loses
Federer        0
Nadal          1
Nadal serves and loses
Federer        1
Nadal          1
Nadal serves and loses
Federer        2
Nadal          1
Federer serves and loses
Federer        2
Nadal          2
Federer serves and loses
Federer        2
Nadal          3
Nadal serves and wins
Federer        2
Nadal          4
Nadal serves and loses
Federer        3
Nadal          4
Federer serves and wins
Federer        4
Nadal          4
Federer serves and wins
Federer        5
Nadal          4
Nadal serves and wins
Federer        5
Nadal          5
Nadal serves and wins
Federer        5
Nadal          6
Federer serves and wins
Federer        6
Nadal          6
Federer serves and loses
Federer        6
Nadal          7
Nadal serves and wins
Federer        6
Nadal          8
Nadal wins TieBreakGame

TieBreakGame starts
Federer serves and wins
Federer        1
Nadal          0
Nadal serves and wins
Federer        1
Nadal          1
Nadal serves and loses
Federer        2
Nadal          1
Federer serves and wins
Federer        3
Nadal          1
Federer serves and wins
Federer        4
Nadal          1
Nadal serves and loses
Federer        5
Nadal          1
Nadal serves and wins
Federer        5
Nadal          2
Federer serves and loses
Federer        5
Nadal          3
Federer serves and loses
Federer        5
Nadal          4
Nadal serves and loses
Federer        6
Nadal          4
Nadal serves and wins
Federer        6
Nadal          5
Federer serves and loses
Federer        6
Nadal          6
Federer serves and loses
Federer        6
Nadal          7
Nadal serves and loses
Federer        7
Nadal          7
Nadal serves and loses
Federer        8
Nadal          7
Federer serves and loses
Federer        8
Nadal          8
Federer serves and wins
Federer        9
Nadal          8
Nadal serves and loses
Federer        10
Nadal          8
Federer wins TieBreakGame

